integration-tests/src/test/java/com/transferwise/tasks/testapp/TaskReschedulingIntTest.java

```diff
@@ -0,0 +1,191 @@
+package com.transferwise.tasks.testapp;
+
+import static com.transferwise.tasks.domain.TaskStatus.WAITING;
+import static org.awaitility.Awaitility.await;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import com.transferwise.common.baseutils.UuidUtils;
+import com.transferwise.tasks.BaseIntTest;
+import com.transferwise.tasks.ITaskDataSerializer;
+import com.transferwise.tasks.ITasksService;
+import com.transferwise.tasks.ITasksService.RescheduleTaskResponse.Result;
+import com.transferwise.tasks.dao.ITaskDao;
+import com.transferwise.tasks.domain.Task;
+import com.transferwise.tasks.domain.TaskStatus;
+import com.transferwise.tasks.management.ITasksManagementService;
+import com.transferwise.tasks.test.ITestTasksService;
+import io.micrometer.core.instrument.Counter;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.UUID;
+import lombok.extern.slf4j.Slf4j;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.EnumSource;
+import org.springframework.beans.factory.annotation.Autowired;
+
+@Slf4j
+public class TaskReschedulingIntTest  extends BaseIntTest {
+
+  @Autowired
+  private ITasksService tasksService;
+  @Autowired
+  private ITestTasksService testTasksService;
+  @Autowired
+  private ITaskDataSerializer taskDataSerializer;
+  @Autowired
+  private ITasksManagementService tasksManagementService;
+  @Autowired
+  private ITaskDao taskDao;
+
+  @BeforeEach
+  void setup() {
+    transactionsHelper.withTransaction().asNew().call(() -> {
+      testTasksService.reset();
+      return null;
+    });
+  }
+
+  @Test
+  void taskCanBeSuccessfullyRescheduled() {
+    testTaskHandlerAdapter.setProcessor(resultRegisteringSyncTaskProcessor);
+    UUID taskId = UuidUtils.generatePrefixCombUuid();
+
+    transactionsHelper.withTransaction().asNew().call(() ->
+        tasksService.addTask(new ITasksService.AddTaskRequest()
+            .setTaskId(taskId)
+            .setData(taskDataSerializer.serialize("I want to be rescheduled"))
+            .setType("test").setRunAfterTime(ZonedDateTime.now().plusHours(1)))
+    );
+
+    await().until(() -> !testTasksService.getWaitingTasks("test", null).isEmpty());
+
+    var task = tasksManagementService.getTasksById(
+        new ITasksManagementService.GetTasksByIdRequest().setTaskIds(List.of(taskId))
```

Comment:
so would something like this be the best practice @xSeagullx? 🤔 TasksService returns all tasks of a certain type -> it seems like an overkill to query everything to get one specific task. But with tasksManagementService we can get task by it's id and get correct version (though is it stable enough to run not on UI but for our use case?)

Reply:
Good Q. Generally — no, you should not fetch all tasks of type to get specific task. Do you need to do it because taskService does not have such method?
Especially, if you already track task ids, there should be a way to get the task by id directly..

Reply:
yeah, taskService doesn't seem to have this capability. Would we need to use tasksManagementService?

Reply:
I think it needs to be added. This won't scale for bigger services (or when your service gets bigger)

Reply:
Actually, nvm me, that's tests